### PR TITLE
v1.14 Backports 2024-01-10

### PR DESCRIPTION
**Pull request by jibi**
* [x] #29237 (@jibi)
 * [x] #29528 (@mhofstetter)
 * [ ] #29971 (@renovate[bot])
 * [x] #29999 (@julianwiedmann)
 * [x] #28910 (@tamilmani1989)
 * [x] #29976 (@cornfeedhobo)
   *  @cornfeedhobo I skipped 88d42303cfda4950715fd46d4fb50f8065217abd as one of the lines with the typo is missing and the other one is already fixed, please double check that we are not missing anything
 * [x] #30017 (@pmcgrath)
   * skipped as the changes were already part of the previous commit 
 * [x] #28947 (@meyskens)
 * [x] #29893 (@giorio94)

PRs skipped due to conflicts:

 * #29986 (@qmonnet)
 * #29400 (@meyskens)

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
29237 29528 29971 29999 28910 29976 30017 28947 29893
```

### CODE REVIEW

**Review by julianwiedmann — APPROVED** [review]
lgtm, thanks!

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,8 @@
 module github.com/cilium/cilium
 
-go 1.20
+go 1.21
+
+toolchain go1.21.1
```
Comment (jibi) [inline go.mod:5]: this made the go linter unhappy

```
/home/runner/work/cilium/cilium/go.mod:5: unknown directive: toolchain
```

**Review by mhofstetter — APPROVED** [review]
thanks!

**Review by tklauser — APPROVED** [review]
My change looks good, thanks Jibi!

**Review by giorio94 — APPROVED** [review]
My commit looks good, thanks!

**Comment by jibi** [comment]
@tklauser it looks like https://github.com/cilium/cilium/pull/27427 is breaking the build:
```
Error: cmd/state.go:205:58: ep.GetK8sNamespaceAndCEPName undefined (type *"github.com/cilium/cilium/pkg/endpoint".Endpoint has no field or method GetK8sNamespaceAndCEPName)
Error: cmd/state.go:432:30: ep.GetK8sNamespaceAndCEPName undefined (type *"github.com/cilium/cilium/pkg/endpoint".Endpoint has no field or method GetK8sNamespaceAndCEPName)
```
that function is defined in ced01f656f0, but I'm not sure we want to backport the full PR?

**Comment by tklauser** [comment]
> @tklauser it looks like #27427 is breaking the build:
> 
> ```
> Error: cmd/state.go:205:58: ep.GetK8sNamespaceAndCEPName undefined (type *"github.com/cilium/cilium/pkg/endpoint".Endpoint has no field or method GetK8sNamespaceAndCEPName)
> Error: cmd/state.go:432:30: ep.GetK8sNamespaceAndCEPName undefined (type *"github.com/cilium/cilium/pkg/endpoint".Endpoint has no field or method GetK8sNamespaceAndCEPName)
> ```
> 
> that function is defined in [ced01f6](https://github.com/cilium/cilium/commit/ced01f656f03149e58065ad1dcbefbcb833009cd), but I'm not sure we want to backport the full PR?

@jibi Sorry about that. That change was already skipped in a previous 1.14 backport attempt in https://github.com/cilium/cilium/pull/30080 due to that dependency. But it seems I forgot to remove the backport labels. Will do so now. Could you please drop the respective changes from this PR.

**Comment by jibi** [comment]
/test-backport-1.14

**Comment by jibi** [comment]
/test-backport-1.14

**Review by viktor-kurchenko — APPROVED** [review]

**Review by meyskens — APPROVED** [review]
LGTM

**Comment by meyskens** [comment]
Any help needed with https://github.com/cilium/cilium/pull/29400?

**Comment by jibi** [comment]
> Any help needed with #29400?

yes please, I got some conflicts in the unit tests that I wasn't sure how to tackle (I think `fakeIPCache` got replaced in main with `fakeNodeIDHandler` and the `deleted` field is missing from the `CertificateRotationEvent` struct)

**Review by squeed — APPROVED** [review]

**Pull request merged**